### ISSUE TITLE
Source Freshness is not implemented

### ISSUE THREAD

**Issue by wrborigin** [starburstdata/dbt-trino#28]
When I add source freshness configuration and run `dbt source freshness` with dbt-trino, I get

```
Runtime Error in source mv_mantis_mfr (models/staging/sources.yml)
  ERROR: current_timestamp macro not implemented for adapter trino
```

**Comment by findinpath** [comment]
This should be solved pretty soon with the migration to v1.0.0


See https://github.com/starburstdata/dbt-trino/commit/8df0f802254150aa3458b731f7c5db74097349af

**Comment by hovaesco** [comment]
@wrborigin `dbt-trino v1.0.0` has been released

**Comment by wrborigin** [comment]
I tested it with dbt-trino 1.0, and it seems to have some problem.

```yml
sources:
  - name: redacted
    database: redacted
    schema: redacted
    loaded_at_field: approval_date
```
```
trino> describe "redacted"."redacted"."redacted" ;
         Column          |     Type     |     Extra     |    Comment   
-------------------------+--------------+---------------+---------------------------
 approval_date           | timestamp(3) |               | 
```
```
17:02:40  Database Error in source mv_mantis_mfr (models/staging/sources.yml)
17:02:40    Expected a timestamp value when querying field 'approval_date' of table "redacted"."redacted"."redacted" but received value of type 'str' instead
```

Any idea what might be wrong?

Even adding `"cast(approval_date as timestamp)"` to the loaded_at_field doesn't help.

**Comment by findinpath** [comment]
Please visit the trino web admin UI and paste the full stack trace of the problem.

**Comment by wrborigin** [comment]
Trino UI shows a **successful** query of

```sql
select
      max(approval_date) as max_loaded_at,
      CURRENT_TIMESTAMP as snapshotted_at
    from "redacted"."redacted"."redacted"
```

If I `prepare` and `describe output` of that query, I get

```
Column Name   |Catalog|Schema|Table|Type                       |Type Size|Aliased|
--------------+-------+------+-----+---------------------------+---------+-------+
max_loaded_at |       |      |     |timestamp(3)               |        8|true   |
snapshotted_at|       |      |     |timestamp(3) with time zone|        8|true   |
```

so Trino definitely return timestamp.

I remembered [I was looking into this for dbt-presto](https://github.com/dbt-labs/dbt-presto/issues/55) sometime in summer, and the problem was that while Trino returns timestamp, the Python driver somehow makes it a string, and that's what dbt is complaning about.

**Comment by findinpath** [comment]
```
Database Error in source stg_orders (models/staging/stg_orders.yml)
  Expected a timestamp value when querying field 'CAST(order_date AS timestamp)' of table "memory"."default"."stg_orders" but received value of type 'str' instead
Traceback (most recent call last):
  File "/Users/marius/Development/starburst/dbt-trino-test/dbt-trino-jaffle-shop-freshness/venv/lib/python3.8/site-packages/dbt/task/base.py", line 361, in safe_run
    result = self.compile_and_execute(manifest, ctx)
  File "/Users/marius/Development/starburst/dbt-trino-test/dbt-trino-jaffle-shop-freshness/venv/lib/python3.8/site-packages/dbt/task/base.py", line 314, in compile_and_execute
    result = self.run(ctx.node, manifest)
  File "/Users/marius/Development/starburst/dbt-trino-test/dbt-trino-jaffle-shop-freshness/venv/lib/python3.8/site-packages/dbt/task/base.py", line 403, in run
    return self.execute(compiled_node, manifest)
  File "/Users/marius/Development/starburst/dbt-trino-test/dbt-trino-jaffle-shop-freshness/venv/lib/python3.8/site-packages/dbt/task/freshness.py", line 149, in execute
    freshness = self.adapter.calculate_freshness(
  File "/Users/marius/Development/starburst/dbt-trino-test/dbt-trino-jaffle-shop-freshness/venv/lib/python3.8/site-packages/dbt/adapters/base/impl.py", line 1094, in calculate_freshness
    max_loaded_at = _utc(table[0][0], source, loaded_at_field)
  File "/Users/marius/Development/starburst/dbt-trino-test/dbt-trino-jaffle-shop-freshness/venv/lib/python3.8/site-packages/dbt/adapters/base/impl.py", line 93, in _utc
    raise raise_database_error(
  File "/Users/marius/Development/starburst/dbt-trino-test/dbt-trino-jaffle-shop-freshness/venv/lib/python3.8/site-packages/dbt/exceptions.py", line 465, in raise_database_error
    raise DatabaseException(msg, node)
dbt.exceptions.DatabaseException: Database Error in source stg_orders (models/staging/stg_orders.yml)
  Expected a timestamp value when querying field 'CAST(order_date AS timestamp)' of table "memory"."default"."stg_orders" but received value of type 'str' instead
```

This is the stacktrace of the `dbt source freshness` command when it gets applied to a model containing freshness checks.

**Comment by findinpath** [comment]
Related issue:
https://github.com/trinodb/trino-python-client/issues/32

**Comment by mdesmet** [comment]
Should be fixed by #61 as we use `experimental_python_types` from trino-python-client.

**Comment by hovaesco** [comment]
Resolved by https://github.com/starburstdata/dbt-trino/pull/61
